lib/internal/modules/helpers.js

```diff
@@ -288,14 +289,30 @@ function addBuiltinLibsToObject(object, dummyModuleName) {
 }
 
 /**
- * If a referrer is an URL instance or absolute path, convert it into an URL string.
- * @param {string | URL} referrer
+ * Normalize the referrer name as a URL.
+ * If it's an absolute path or a file:// it's normalized as a file:// URL.
+ * Otherwise it's returned as undefined;
+ * @param {string | null | undefined | false | any } referrerName
+ * @returns {string | undefined}
  */
-function normalizeReferrerURL(referrer) {
-  if (typeof referrer === 'string' && path.isAbsolute(referrer)) {
-    return pathToFileURL(referrer).href;
+function normalizeReferrerURL(referrerName) {
+  if (typeof referrerName !== 'string') {
+    return undefined;
   }
-  return new URL(referrer).href;
+
+  if (StringPrototypeStartsWith(referrerName, 'file://')) {
```

Comment:
We can’t do this. I proposed this in https://github.com/nodejs/node/issues/48994 but it was shot down because you can, in fact, have a path that begins with `file://`; and it’s actually not all that uncommon, when people mirror URLs in a local filesystem. The logic needs to be “if it’s an instance of the URL class, parse as URL; else parse strings as paths.” cc @aduh95

Reply:
We do not accept URL in the API. This is the current behavior and disallowing file:// paths is semver major.

Reply:
Updated to simply return the string as-is instead of parsing it at all when it looks like a URL. However the default loader treat a module with a file:// URL string name is up to the default loader, we are only passing the name over.